### PR TITLE
Custom renderer for exceptions

### PR DESCRIPTION
**Pull request by Giorgi**
Implementation of custom renderer for exceptions for issue #3519

### CODE REVIEW

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs [hunk 1]
```diff
@@ -285,11 +296,29 @@ private void AppendException(Exception currentException, IEnumerable<ExceptionRe
...
                 var currentRenderFunction = _renderingfunctions[renderingFormat];
                 currentRenderFunction(builder, currentException);
+
                 if (builder.Length == beforeRenderLength && builder.Length != orgLength)
                 {
                     builder.Length = orgLength;
                 }
             }
+
+            if (_customRenderingfunctions.TryGetValue(currentException.GetType(), out var renderer))
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: Is this isn't a issue for subclasses?

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs [hunk 2]
```diff
@@ -177,6 +178,16 @@ public List<ExceptionRenderingFormat> InnerFormats
...
         }
 
+        /// <summary>
+        /// Registers a custom exception renderer type
+        /// </summary>
+        /// <typeparam name="T">Type of the exception</typeparam>
+        /// <param name="renderer">The renderer for converting exception to string</param>
+        public static void RegisterExceptionRenderer<T>(Func<T, string> renderer) where T : Exception
+        {
+            _customRenderingfunctions.Add(typeof(T), exception => renderer((T) exception));
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: `_customRenderingfunctions[type] = ` is maybe better (no exception when duplicate)

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs [hunk 3]
```diff
@@ -285,11 +296,29 @@ private void AppendException(Exception currentException, IEnumerable<ExceptionRe
...
+                    builder.Append(Separator);
+                    var render = renderer(currentException);
+                    builder.Append(render);
+                }
+                catch (Exception exception)
+                {
+                    var message = $"Exception in {typeof(ExceptionLayoutRenderer).FullName}.AppendException(): {exception.GetType().FullName}.";
+                    builder.Append("NLog message: ");
+                    builder.Append(message);
+                    InternalLogger.Warn(exception, message);
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: Only log to internallog here?

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs (hunk 1 above)
Comment (Giorgi) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: @304NotModified Do you mean that if there is no custom renderer for this type it should use a renderer for base type?

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs (hunk 3 above)
Comment (Giorgi) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: Probably, I copied it from `AppendMessage` in the same class

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs (hunk 2 above)
Comment (Giorgi) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: Done

**Comment by snakefoot** [comment]
One could also consider making use of ObjectReflectionCache, so it will automatically output any extra properties on the Exception-object (Without object-reference-recursion by default).

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: If you have the classes like this,

    Exception <- MyExceptionBase <- MyException

And registering for MyExceptionBase, it won't work for MyException?

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs (hunk 3 above)
Comment (304NotModified) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: Have to check this :)

**Comment by snakefoot** [comment]
Making use of ObjectReflectionCache would also allow "automatic" support of generic exceptions and base-classes:

https://stackoverflow.com/questions/57147623/nlog-capture-faultexceptiontdetail-in-log

**Comment by 304NotModified** [comment]
Cool idea!

**Comment by Giorgi** [comment]
The disadvantage of `ObjectReflectionCache` is that it handles only properties. For example in case of `WebException` you often need to read the stream by calling `WebException.Response.GetResponseStream` method which you can't do with ObjectReflectionCache

**Comment by snakefoot** [comment]
> The disadvantage of ObjectReflectionCache is that it handles only properties. 

I you take a closer look then you will see it is possible to register custom-methods that can extract properties from an object of a certain type.

But the code is currently not public accessible, because I waited for a proper use-case to arrive (maybe this is the one?), but I also try to avoid static registration methods so not that keen on opening the door until we have proper DependencyInjection (DI) available in NLog. 

Maybe the registration of type-reflection-helpers should just be registration of a generic interface like ex. `IPropertyReflection<T>`. Then one could just register the interface in the global DI-container instead.

But I will try to show how to use `ObjectReflectionCache` then we can continue the discussion of how to register custom methods for extracting custom values. But my time is limited, so not able to work full time on NLog. Lots of stuff is missing for NLog 5.0

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs (hunk 3 above)
Comment (304NotModified) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: Only internal log please, thanks!

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: You could use `IsAssignableFrom` for this, unfortunately it's not working a dictionary then. (or you should write a fancy comparer?)

**Comment by 304NotModified** [comment]
> I you take a closer look then you will see it is possible to register custom-methods that can extract properties from an object of a certain type.

I think this is nice, but maybe to heavy for now?

**Comment by 304NotModified** [comment]
maybe do both, but I think objectpath in the exceptionrenderer isn't a bad idea (like that one in https://github.com/NLog/NLog/wiki/EventProperties-Layout-Renderer).

For those properties you don't need c# code for configuring

**Comment by snakefoot** [comment]
I'm not thinking objectPath but more like format=@. But instead of json. Then it adds extra properties not included in default tostring output. Ex. format=tostring,data,properties. Where properties means starting the reflection cache but very restricted so it will not navigate a DbContext (will also skip Data, StackTrace, etc)


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


maybe do both, but I think objectpath in the exceptionrenderer isn't a bad idea (like that one in https://github.com/NLog/NLog/wiki/EventProperties-Layout-Renderer).

For those properties you don't need c# code for configuring

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: I mean IsInstanceOfType

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs (hunk 3 above)
Comment (Giorgi) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: @304NotModified Done

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs (hunk 1 above)
Comment (Giorgi) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: @304NotModified I just loop thought the base types if there is no renderer for current exception type

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs [hunk 4]
```diff
@@ -303,19 +303,28 @@ private void AppendException(Exception currentException, IEnumerable<ExceptionRe
                 }
             }
 
-            if (_customRenderingfunctions.TryGetValue(currentException.GetType(), out var renderer))
+            var exceptionType = currentException.GetType();
+            while (exceptionType != typeof(Exception))
```
Comment (snakefoot) [inline src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: Please check if `_customRenderingfunctions.Count > 0` before starting all this logic.

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs [hunk 5]
```diff
@@ -285,11 +296,37 @@ private void AppendException(Exception currentException, IEnumerable<ExceptionRe
...
+                    {
+                        var message = $"Exception in {typeof(ExceptionLayoutRenderer).FullName}.AppendException(): {exception.GetType().FullName}. Custom renderer for {currentException.GetType().Name} threw an exception";
+
+                        InternalLogger.Warn(exception, message);
+                    }
+                    break;
+                }
+                else
+                {
+                    exceptionType = exceptionType.BaseType;
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: I think this doens't work for interface ;)

maybe a some test cases first?

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs (hunk 4 above)
Comment (snakefoot) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: How do you use registered `_customRenderingfunctions` alone without getting the output from the configured `format`-property ? (Default = `message`).

Comment (Giorgi) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: @snakefoot You can't do that currently. I don't think anyone would want to log specific properties and not message. I see custom renderer as an additional detail provider and if everyone needs a message NLog shouldn't force users to include message in every custom renderer

Comment (snakefoot) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: I was trying to nudge you in the direction that you should add new enum value to ExceptionRenderingFormat

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs (hunk 5 above)
Comment (Giorgi) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: @304NotModified When would exceptionType ever be an interface? `currentException.GetType()` will always return an exception type representing a class so accessing BaseType hierarchically should never be an interface type.

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs (hunk 4 above)
Comment (304NotModified) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: >  I don't think anyone would want to log specific properties and not message.

one use case is to log both, but not in the same (database) column/field.

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs (hunk 5 above)
Comment (304NotModified) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: ` public static void RegisterExceptionRenderer<IMyExceptionInterface>(...)`

Comment (Giorgi) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: RegisterExceptionRenderer is limited to Exception subclasses at the moment:

`public static void RegisterExceptionRenderer<T>(Func<T, string> renderer) where T : Exception`

Comment (304NotModified) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: not sure if that's a good idea. We could enumerate the list and use IsInstanceOfType? Then it also works if someone nice abstractions. :)

Comment (Giorgi) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: If someone registers a renderer for `MyExceptionBase` and than registers for `MyException` `IsInstanceOfType` will return true for both of them

Comment (Giorgi) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: @304NotModified Also, if the type implements two interfaces and someone registers custom renderer for both of them which one should be used ?

**Comment by snakefoot** [comment]
Instead of registering custom formatter on the exception-layout-renderer. Then one could also consider injecting the custom rendering using the IValueFormatter-interface.

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs (hunk 5 above)
Comment (304NotModified) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs]: Both or closest path.

**Comment by 304NotModified** [comment]
I think this PR isn't going in the right direction as it doesn't support interfaces. I think we should do the design before implementing. 

Another option is to implement this is your own code (derive from the layout renderer) and maybe change something in NLog for the integration

**Comment by Giorgi** [comment]
@304NotModified I can add support for interfaces if supporting them is such a top priority.

**Comment by 304NotModified** [comment]
@Giorgi  I think without interface, this will be a strange addition.

But I also added a new layout renderer, just for rendering properties. You could see it here, https://github.com/NLog/NLog/pull/3531 (needs review)

That will handle properties on more levels than ${exception} and don't need any configuration in code :)

**Comment by 304NotModified** [comment]
Closing this due to inactivity. Feel free to open a new PR in the future!

**Pull request closed without merging**